### PR TITLE
Add pagination to browse page

### PR DESCRIPTION
**Pull request by Zetaphor**

### CODE REVIEW

**Comment by avaer** [comment]
This looks good and does what it says!

Is it possible to dynamically load each page without waiting on a loading screen for everything to download first?

In `master` we dynamically insert the elements as they come in -- we could probably do the same thing here, on a per-page basis.

In the long term we'd probably cache this load and make it instant coming from the server, but for now I think it's better to see the results of the query as fast as possible even if there is pop-in.

**Comment by Zetaphor** [comment]
> Is it possible to dynamically load each page without waiting on a loading screen for everything to download first?

Yes, I could do this on the initial load

> In master we dynamically insert the elements as they come in -- we could probably do the same thing here, on a per-page basis.

> In the long term we'd probably cache this load and make it instant coming from the server, but for now I think it's better to see the results of the query as fast as possible even if there is pop-in.

Currently the query is getting all of the results from the server, and I'm saving that data in `pkgResults` so that I can eventually add searching against the local array. Is there a way to only request a subset of results from the server? I admittedly didn't look into the `web3.js` library at all.

If yes then I could take this further and implement the pagination on the query and save a lot of load time

**Comment by avaer** [comment]
> Is there a way to only request a subset of results from the server? 

Yes. Here is where we iterate over the packages list as we fetch: https://github.com/webaverse/xrpackage/blob/95adbbe40aca01e67adc4620646286f45310f735/browse.html#L128

Packages start at id `1` and go up to `nonce`; that loop can be a subset.

> eventually add searching against the local array

That makes sense, though I think for a reasonably sized list that quickly starts to make more sense on the server side rather than the client side.

**Comment by Zetaphor** [comment]
😳 You weren't supposed to see that force push lol, for some reason my local git config name is screwed up 🤷‍♂️ 

I've implemented the requested changes. The results are now rendered as they're loaded and the server is only queried for the current pages subset of objects

**Comment by avaer** [comment]
For some reason I was able to get 11 items on the first page.

Screen Shot 2020-04-13 at 12 41 14 AM

The second page starts at 10 and duplicates items from the first page:

Screen Shot 2020-04-13 at 12 41 48 AM

**Comment by Zetaphor** [comment]
There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors

All set

File: browse.html [hunk 1]
```diff
@@ -51,56 +70,115 @@
...
-        <a href="https://ipfs.exokit.org/ipfs/${modelHash}.glb" class="cardAction"><span>Download model</span><i class="fa fa-chevron-right"></i></a>
-        <a href="https://${network}.opensea.io/assets/${address}/${i}" class="cardAction"><span>Opensea</span><i class="fa fa-chevron-right"></i></a>
-      </div>
-    `;
-    const input = div.querySelector('input');
-    input.addEventListener('click', e => {
-      input.select();
-    });
-    grid.appendChild(div);
+  nonce, totalObjects = parseInt(nonce, 10);
```
Comment (avaer) [inline browse.html:150]: I don't understand this line; what is the `nonce,` for?

Comment (Zetaphor) [reply browse.html]: That was leftover from the previous implementation, fixed in 1d7b945

**Comment by Zetaphor** [comment]
> Actually I lied, page 2 looks like it duplicates item 10 from the previous page:

Resolved in b361376

**Comment by avaer** [comment]
For future, it would also be good to have the pagination loaded by query string, so a specific page can be linked to when browsing.

**Pull request merged**